### PR TITLE
Add type cast support

### LINKED ISSUE #14406: Need to invoke init function during test setup
Currently __init_() method get generated, however, it is needed to be invoked before running the test cases. Otherwise, global level variables, and user defined types will not get initialized. e.g.:

ballerina test function:
```ballerina
int globalVar = 7;

function getGlobalVar() returns int {
    return globalVar;
}
```
Java test case:
```java
    @Test
    public void getGlobalVar() {
        BValue[] returns = BRunUtil.invoke(compileResult, "getGlobalVar");
        Assert.assertEquals(((BInteger) returns[0]).intValue(), 7);
    }
```

### PR DESCRIPTION
**Pull request by SupunS**
Resolves #14406

### CODE REVIEW

File: compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_method_gen.bal [hunk 1]
```diff
@@ -133,6 +137,8 @@ function generateMethod(bir:Function func, jvm:ClassWriter cw) {
                 } else if (inst.kind == "ARRAY_LOAD") {
                     instGen.generateArrayValueLoad(inst);
                 }
+            } else if (inst is bir:TypeCast) {
```
Comment (vinok88) [inline compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_method_gen.bal]: I think we do not need this else if.

File: compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_cast_gen.bal [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    if (sourceType is bir:BTypeAny || sourceType is bir:BTypeAnyData) {
+        mv.visitTypeInsn(CHECKCAST, targetClass);
+    } else {
+        checkCast(mv, sourceType, targetType);
+        return;
+    }
+}
+
+function checkCast(jvm:MethodVisitor mv, bir:BType sourceType, bir:BType targetType) {
+    // TODO: invoke checkIsType() native function
```
Comment (Kishanthan) [inline compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_cast_gen.bal]: shall we call the static function that does the cast at runtime in JVM here directly ?

File: compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_instruction_gen.bal [hunk 3]
```diff
@@ -441,25 +441,38 @@ type InstructionGenerator object {
             }
         }
     }
+
+    function generateCastIns(bir:TypeCast typeCastIns) {
+        // load source value
+        int sourceIndex = self.getJVMIndexOfVarRef(typeCastIns.rhsOp.variableDcl);
```
Comment (Kishanthan) [inline compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_instruction_gen.bal]: this index loading is not needed

File: compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_instruction_gen.bal [hunk 4]
```diff
@@ -441,25 +441,38 @@ type InstructionGenerator object {
...
+        self.generateVarLoad(typeCastIns.rhsOp.variableDcl);
+
+        generateCast(self.mv, typeCastIns.rhsOp.typeValue, typeCastIns.lhsOp.typeValue);
+
+        int targetIndex = self.getJVMIndexOfVarRef(typeCastIns.lhsOp.variableDcl);
```
Comment (Kishanthan) [inline compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_instruction_gen.bal]: same here, this index loading is not needed

File: tests/ballerina-test-utils/src/main/java/org/ballerinalang/test/util/BCompileUtil.java [hunk 5]
```diff
@@ -72,6 +76,7 @@
 public class BCompileUtil {
 
     private static Path resourceDir = Paths.get("src/test/resources").toAbsolutePath();
+    private static final JBallerinaInMemoryClassLoader classLoader = new JBallerinaInMemoryClassLoader();
```
Comment (Kishanthan) [inline tests/ballerina-test-utils/src/main/java/org/ballerinalang/test/util/BCompileUtil.java]: shall we move this classLoader instance within the compileOnJBallerina method it self? Also shall we make the byte[] as the constructor argument for JBallerinaInMemoryClassLoader?

File: tests/ballerina-test-utils/src/main/java/org/ballerinalang/test/util/BCompileUtil.java [hunk 6]
```diff
@@ -409,8 +414,20 @@ private static CompileResult compileOnJBallerina(String sourceFilePath) {
 
         BLangPackage bLangPackage = (BLangPackage) compileResult.getAST();
         byte[] compiledJar = JVMCodeGen.generateJarBinary(bLangPackage, context, packageName);
-        compileResult.setCompiledJarFile(compiledJar);
         compileResult.setEntryClassName(FileUtils.cleanupFileExtension(packageName));
+        classLoader.setClassContent(compiledJar);
+        Class<?> clazz = classLoader.loadClass(compileResult.getEntryClassName());
+
+        // invoke the init function
+        String funcName = "__init_";
```
Comment (Kishanthan) [inline tests/ballerina-test-utils/src/main/java/org/ballerinalang/test/util/BCompileUtil.java]: done we need orgname and modulename here?

File: tests/ballerina-test-utils/src/main/java/org/ballerinalang/test/util/BCompileUtil.java [hunk 7]
```diff
@@ -409,8 +414,20 @@ private static CompileResult compileOnJBallerina(String sourceFilePath) {
 
         BLangPackage bLangPackage = (BLangPackage) compileResult.getAST();
         byte[] compiledJar = JVMCodeGen.generateJarBinary(bLangPackage, context, packageName);
-        compileResult.setCompiledJarFile(compiledJar);
         compileResult.setEntryClassName(FileUtils.cleanupFileExtension(packageName));
```
Comment (Kishanthan) [inline tests/ballerina-test-utils/src/main/java/org/ballerinalang/test/util/BCompileUtil.java]: setEntryClassName and getEntryClassName methods are not needed anymore right? they are used within this method itself

File: compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_instruction_gen.bal (hunk 4 above)
Comment (Kishanthan) [reply compiler/ballerina-backend-jvm/src/main/ballerina/compiler_backend_jvm/jvm_instruction_gen.bal]: lets fix everywhere for the same

**Pull request merged**